ops/main.py

```diff
@@ -384,12 +387,28 @@ def main(charm_class: Type[ops.charm.CharmBase],
     dispatcher = _Dispatcher(charm_dir)
     dispatcher.run_any_legacy_hook()
 
-    metadata = (charm_dir / 'metadata.yaml').read_text()
-    actions_meta = charm_dir / 'actions.yaml'
-    if actions_meta.exists():
-        actions_metadata = actions_meta.read_text()
-    else:
-        actions_metadata = None
+    # Check charmcraft.yaml first and see if it has the metadata data in it.
+    charmcraft_meta = charm_dir / "charmcraft.yaml"
+    if charmcraft_meta.exists():
+        charmcraft_yaml = charmcraft_meta.read_text()
+        charmcraft_metadata: dict[str, Any] = yaml.safe_load(charmcraft_yaml)
+        if any(
+            meta_key in charmcraft_metadata for meta_key in ("name", "summary", "description")
+        ):
+            # Unrelated keys in the charmcraft.yaml file will be ignored.
+            metadata = charmcraft_yaml
+
+        # This could be None if the charmcraft.yaml not include the actions key.
+        # Should fall back to the actions.yaml file.
+        actions_metadata = charmcraft_metadata.get("actions")
+
+    if not metadata:
+        metadata = (charm_dir / "metadata.yaml").read_text()
```

Comment:
In charmcraft's behaviour, is there any merging? Does charmcraft raise an error if the charm has metadata in both `charmcraft.yaml` and `metadata.yaml`? Should we raise an error too, to catch it early?

Reply:
charmcraft will raise error when it found metadata in `charmcraft.yaml` but `metadata.yaml` also exists. `actions` and `config` also have the same behavior.